### ISSUE TITLE
Can't get rid of 'Name source files using `lowercase_with_underscores`.dart(file_names)'

### ISSUE THREAD

**Issue by Asperson** [Dart-Code/Dart-Code#3236]
Hi all, love what you doing.

Yesterday renamed one of my widgets from **tracking_sheet.dart** to **Measure_sheet.dart**. Of course got warning, immediately renamed it to **measure_sheet.dart**, but warning didn't go away.

Renamed it to **Kmeasure_sheet.dart**, then back to **measure_sheet.dart**, the warning appeared again:
![warning](https://user-images.githubusercontent.com/17258058/112788756-8850fd00-9029-11eb-923d-6defc9171197.png)

I tried also removing file from the project and re-adding - no luck.
Please help how can I kill the warning.

Best regards,
Dmitry

**Comment by DanTup** [comment]
I'm not able to reproduce this, the error disappears once I rename the file back. In the logs I can see the server sending an update to clear them.

Could you try capturing a log?

- Open your project and wait for analysis to finish
- Run the **Dart: Capture Analysis Server Logs** command from the command palette
- Rename the file to trigger the error
- Rename the file back and wait a few seconds
- Click the **Cancel** button on the logging notification to stop logging and open the log file

Note: The log file may include parts of your source code so it would be better to reproduce on a small sample project, but otherwise be sure to check the log for anything sensitive before sharing.

Thanks!

**Comment by Asperson** [comment]
I was able to reproduce it on a small project, copying and reducing existing measure_sheet.dart.
One note - I renamed file in the vscode using F2 in the project tree, not externally.
After vscode loaded the file looked correctly but then if I click another file and return back to this one it become reported having incorrect cases in its name.

[Dart-Code-Log-2021-02-01 11-38-48.txt](https://github.com/Dart-Code/Dart-Code/files/6222887/Dart-Code-Log-2021-02-01.11-38-48.txt)

vscode v1.54.3 (user setup)
dart-code v3.20.1

**Comment by DanTup** [comment]
Strange - I renamed the same way, though no luck. I tried on Windows too but it seemed fine (I tried both changing only the casing, and also changing the casing and adding more letters at the same time).

Can you confirm exactly what sort of rename you're making? If I can't repro, I'll see if I can trace through the server how this is normally handled to see if I can come up with any theories for why it wouldn't be happening for you.

Thanks!

**Comment by Asperson** [comment]
I think I found the way to reproduce bug:

1. create flutter new app with 'flutter create'
2. copy attached files to the new project, pubspec.yaml, analysis_options.yaml and aa_aa.dart, last one to lib folder.
3. get packages
3. set MyApp.home to AaWidget()
4. close vscode and open it again
5. rename aa_aa.dart in vscode by F2 to Bb_aa.dart, say 'yes' to allow extension Dart to make refactoring changes
6. rename the same way to bb_aa.dart

After these steps on my machine (windows 10) all content in bb_aa.dart undercrossed with blue line and Problem window shows a case problem for Bb_aa.dart file

[add2project.zip](https://github.com/Dart-Code/Dart-Code/files/6225999/add2project.zip)

**Comment by cto-leaps** [comment]
I have the same issue on Mac OS, but the error is only shown whenever I make any change to the file and remains until I save the file again. Then it disappears. Makes it hard to edit a file...

**Comment by DanTup** [comment]
> but the error is only shown whenever I make any change to the file and remains until I save the file again

Aha! I couldn't repro with the instructions above, but when I modified the file the error appeared again. I'll do some digging. Thanks!

**Comment by DanTup** [comment]
Also - after restarting VS Code I could repro it the original way - just renaming the file to Bb and then back to bb - the error remained.

**Comment by DanTup** [comment]
This feels like a VS Code bug. When I change the casing of a file, then reload VS Code, then close and re-open the file, VS Code still keeps giving me the old case - even though it shows the new casing in its UI.

I've opened https://github.com/microsoft/vscode/issues/121106 to get some clarification (but it's not clear based on https://github.com/microsoft/vscode/issues/102627 whether this is intended or not).

**Comment by DanTup** [comment]
Unfortunately the VS Code issue has been resolved as "by design" and it's not clear there's a way to fix this. VS Code shows a file opened with the old casing even though that's not actually the case.

**Comment by MooNag** [comment]
Hey everyone I had the same issue but I think I found a temporary workaround for it. It basically involves clearing any "cache" and "state" files for vs code. This method will reset your workspace (I believe that includes opened folders, window layout, etc). It should not touch extensions or settings.

Steps for windows (mac and linux may be similar but different locations):
Open your local roaming folder:
`%appdata%/Code` (C:\Users\*USERNAME*\AppData\Roaming\Code)

In this folder delete `Cache`, `Cached Data`, `CachedExtensions`, `Code Cache`, and `User/workspaceStorage`

You can avoid removing all of your workspaces by looking into the workspace folder and finding the one associated with your project and only removing that (I have not tested that).

This should work around the issue. You will need to do this each time it comes up or if you are refactoring many files you can do it at the end of the refactor.

**Comment by ScottishRoss** [comment]
@MooNag's workaround works.

**Comment by DanTup** [comment]
I have been trying to find a reasonable fix for this, but so far nothing has worked. I was planning to use the close/open events on a file rename (https://github.com/microsoft/language-server-protocol/issues/1203) to try and start mapping the casing of the file ourselves, but in VS Code those doesn't seem to fire for this type of rename.

I've added some comments on the LSP cases, although I'm not hopeful for a change, so may have to find some other way to handle this.

**Comment by MooNag** [comment]
> @MooNag's workaround works.
> Edit: It didn't work for long. After VS Code fully reinitialised, the error returned.
> 
> What's really weird about this issue is that it's only affecting one file, not the others.

Did you make sure to delete everything and do it while vs code is closed? It fully worked for my project even after many restarts. I also had an issue with github but managed to resolve it by copying out all files making a commit removing them (it will basically break your commit history; so if you wanted to make a new repository then is when you want to do that) then renaming all of the files after copying them back in and commit the "new" files.

**Comment by DanTup** [comment]
Actually, my failed experiment probably only prevents this from working within a session. If you were to reload VS Code, I think it would fix that - so there's probably still value in it even without getting the right events. I'll have another shot at this soon. Restarting VS Code is certainly a better fix than having to delete its internal caches to get rid of the message.

**Comment by MooNag** [comment]
@DanTup can the plugin issue a VS Code restart? If you can detect the issue; then just explain the issue; and ask if they want to restart to fix it. For multiple file name changes have an option to wait until you are done.

**Comment by DanTup** [comment]
It could, but it's quite agressive (it terminates the extension host which will tear down all other extensions and restart them) so we try to avoid it (we used to do it in the past and the Code team weren't particularly happy, so we changed our own restarts to just recycle our own services instead).

I'll try and get a fix first though and we can go from there. It might be that just closing the file and re-opening it will also work - it depends whether VS Code sends the close/open event or not (sometimes it does not for a few minutes after closing, for performance reasons in case you re-open it soon after).

**Comment by DanTup** [comment]
My fix didn't work due to some other issues, though I'm not yet sure whether they'll need fixing in the analysis server or VS Code. I've opened https://github.com/microsoft/vscode/issues/128198 for some discussion, so for now this is blocked on some clarification there.

**Comment by ArunEdathadan** [comment]
Issue fixed, check this link.  Don't use any uppercase character in the file name.

https://github.com/dart-lang/linter/issues/18#issue-57127039

**Comment by DanTup** [comment]
@ArunEdathadan this issue is related to that lint but it's not fixed. The problem is that after fixing the lint, you may continue to see the warning reported. The problem is that VS Code keeps using the old casing in the API (and provides that incorrect casing to extensions, and persists it across restarts) which makes it difficult for the server to know whether the problem is really fixed or not (versus you having an unsaved file open in VS Code with the old casing).

**Comment by walshlogic** [comment]
For what it's worth I'm having the same issue. This is new as I've used vscode/flutter extensively for the past 9 months and just began after last update a couple days ago. I've tried the fixes listed above with no luck. Only happening when creating new dart files and does not happen on every new file.
Could not find a common factor suce has naming conventions, case issues, special characters. It's fickle but not sure what the key is.
I'll be following this thread and others in hope of a solution. Having to create new files multiple times with difference names/methods before getting one that works. very, very, :(

**Comment by ArunEdathadan** [comment]
> @ArunEdathadan this issue is related to that lint but it's not fixed. The problem is that after fixing the lint, you may continue to see the warning reported. The problem is that VS Code keeps using the old casing in the API (and provides that incorrect casing to extensions, and persists it across restarts) which makes it difficult for the server to know whether the problem is really fixed or not (versus you having an unsaved file open in VS Code with the old casing).



> @ArunEdathadan this issue is related to that lint but it's not fixed. The problem is that after fixing the lint, you may continue to see the warning reported. The problem is that VS Code keeps using the old casing in the API (and provides that incorrect casing to extensions, and persists it across restarts) which makes it difficult for the server to know whether the problem is really fixed or not (versus you having an unsaved file open in VS Code with the old casing).

Mine was fixed. Not sure why.  Make sure file name and class names are in proper format.

**Comment by DanTup** [comment]
@walshlogic 

>  I've tried the fixes listed above with no luck. Only happening when creating new dart files and does not happen on every new file.

Are you saying you're seeing this on files that have _never_ been cased with uppercase first characters? If so, that sounds like a different issue (and perhaps one that's easier to fix). The issue here is if a file is first created with an uppercase first letter and then renamed. VS Code forever gives us the original casing and trying to handle that is unfortunately quite complicated because we have to account for in-memory files that haven't been saved - eg. it's possible you really did create a new file with that casing and it's not the one we think on disk, and because VS Code (technically incorrectly) uses the platform - instead of the filesystem - to decide when to be case sensitive.

**Comment by DanTup** [comment]
I've pushed a workaround for this until either VS Code or the analysis server can handle it better. There's a slight performance penalty to it so it's _off_ by default and can be enabled with the `dart.normalizeFileCasing` setting.

When the setting is set to `true`, paths sent to the server will be compared to the casing on disk and rewritten if they do not match.

If you use this setting and feel like the performance hit is high, please let me know and I can look at whether it may be possible to cache the results for a little while to improve things.

**Comment by khursheed33** [comment]
It's a naming convention made by flutter team, it restricts that you must follow : 
file_name_subname  
pattern for your files and don't use :
FileNameSubname or fileNameSubname  

so if you are using 2.5.0 you must follow the naming convention otherwise you have to add ignore comment to ignore the naming convention.

You can ignore as Follow: 
// ignore: file_names

**Comment by DanTup** [comment]
@khursheed33 the issue here is that after fixing the filenames, the error did not disappear (because of the way VS Code, Dart-Code and the Dart analysis server all handle case-sensitivity in different - but generally incorrect - ways). The next version of Dart-Code has a setting (`dart.normalizeFileCasing`) that can be enabled that will work around the issue until there's a proper fix (but comes with a small performance penalty).

**Comment by Xananax** [comment]
Just to be clear, if I understand correctly, there is no way to make the error go? Even deleting the VSCode cache doesn't work?

That's very strange. Since it persists across restarts, surely it must be logged somewhere?

**Comment by DanTup** [comment]
> Just to be clear, if I understand correctly, there is no way to make the error go? Even deleting the VSCode cache doesn't work?

In the next release there is a setting `dart.normalizeFileCasing` that will prevent the issue by converting all paths to match their casing on the file-system (although this comes with a small performance penality to do this).

> That's very strange. Since it persists across restarts, surely it must be logged somewhere?

Yep, VS Code stores the filename in its data model across reloads and reuses the original case. The idea was that it didn't want to change the casing of URIs on extensions as they might not handle this correctly, so it always uses the same case. I'm not certain this was a good idea though, since it seems to lead to other issues like this (which is made worse by VS Code incorrectly using the platform to decide whether the filesystem is likely case sensitive, rather than actually checking the filesystem).

I would certainly expect clearing VS Code caches to clear the error, but it's ofcourse not a very good solution.

**Comment by BlarCkat** [comment]
find "analysis_options.yaml", and add this under rules:
  rules:
_file_names: false_ 
wait for about 4 secs
![file names](https://user-images.githubusercontent.com/34066158/134791277-d2fd3a7c-14ce-4e7d-b5a4-74899a3b7e73.png)

**Comment by Leonardogf12** [comment]
Someone managed to solve?

**Comment by DanTup** [comment]
Please see the comment at https://github.com/Dart-Code/Dart-Code/issues/3236#issuecomment-923767560.

There's a beta version of that release available at https://github.com/Dart-Code/Dart-Code/releases/tag/v3.27.0-beta.2.

**Comment by Sin333** [comment]
> find "analysis_options.yaml", and add this under rules:
> rules:
> _file_names: false_
> wait for about 4 secs
> ![file names](https://user-images.githubusercontent.com/34066158/134791277-d2fd3a7c-14ce-4e7d-b5a4-74899a3b7e73.png)

Thanks u so much)

**Comment by easeccy** [comment]
This worked for me:

- Backup the content of the file and remove it.
- Restart VSCode.
- Create the file and paste the code inside.

**Comment by refat47** [comment]
> Hi all, love what you doing.
> 
> Yesterday renamed one of my widgets from **tracking_sheet.dart** to **Measure_sheet.dart**. Of course got warning, immediately renamed it to **measure_sheet.dart**, but warning didn't go away.
> 
> Renamed it to **Kmeasure_sheet.dart**, then back to **measure_sheet.dart**, the warning appeared again: ![warning](https://user-images.githubusercontent.com/17258058/112788756-8850fd00-9029-11eb-923d-6defc9171197.png)
> 
> I tried also removing file from the project and re-adding - no luck. Please help how can I kill the warning.
> 
> Best regards, Dmitry



> 

Just simply edit your .dart file to lower case with _ and it will be resolved

**Comment by hann-1** [comment]
> > find "analysis_options.yaml", and add this under rules:
> > rules:
> > _file_names: false_
> > wait for about 4 secs
> > ![file names](https://user-images.githubusercontent.com/34066158/134791277-d2fd3a7c-14ce-4e7d-b5a4-74899a3b7e73.png)
> 
> Thanks u so much) After all need reload vs code (actual for mac os)

This helps me! Thanks!

**Comment by BlarCkat** [comment]
👍👍

**Comment by Omi231** [comment]
deleting `User/workspaceStorage` inside `(C:\Users*USERNAME*\AppData\Roaming\Code)` worked for me.

**Comment by DanTup** [comment]
@jaidex are you able to reproduce that issue? I can't think of a reason why a mismatched import would trigger this issue (although I can imagine it could cause some other issues while running/debugging). If you can reproduce it, or if the `dart.normalizeFileCasing` setting doesn't prevent it, I'd definitely like to understand why it's failing.

Thanks!

**Comment by Lorenzohidalgo** [comment]
> The version with this fix in actually shipped last week. See https://dartcode.org/releases/v3-27/#editor :-)

@DanTup it looks like this issue has "re-surfaced" at least in our case.

Here are the versions we are working with that show the incorrect warning:
Flutter: v3.3.3
Dart: v2.18.2
VSCode: v1.72.0
Dart Code Extension: v3.50.0

**Comment by DanTup** [comment]
@Lorenzohidalgo do you have the `dart.normalizeFileCasing` setting enabled? If so, can you capture a log as described [in this comment above](https://github.com/Dart-Code/Dart-Code/issues/3236#issuecomment-809219205)? Thanks!

**Comment by Lorenzohidalgo** [comment]
@DanTup sorry, I didn't have that config initially enabled. 
I've reproduced the steps mentioned in the linked comment but could not reproduce our error with either `dart.normalizeFileCasing` set to `selected` or `unselected`.

In our case, this issue happened initially on my colleague's system and "passed" on to me when I pulled the changes. There was no "file name change" happening on my computer and I can only dismiss the warning by adding `// ignore_for_file: file_names` to the files. 

I'll try to reproduce all steps again while capturing the logs. Since the "use case"/issue is not precisely the same, should I share them on this issue or create a new one?

**Comment by DanTup** [comment]
@Lorenzohidalgo do you mean the issue has gone away now and you can no longer reproduce?

If you can reproduce an issue with `normalizeFileCasing` enabled, please do file a new issue. `normalizeFileCasing` is a bit of a hack and not the real solution, but it's all we have right now. Unfortunately the real fix is rather complicated because it requires VS Code, LSP and the Dart server to all agree on how case sensitivity should work, but right now [VS Code is basing its case sensitivity purely on platform](https://github.com/microsoft/vscode/issues/94307) (which is not always correct, and makes it perhaps not a great model for LSP/language servers to follow).

**Comment by ho-lapang** [comment]
![Screenshot_20221009-020636.png](https://user-images.githubusercontent.com/115382488/194808652-952b485c-f849-48a1-978c-55766be08ab1.png)

**Comment by PeterHenter** [comment]
I had the same issue. 

My versions are as follows:
Flutter 3.7.0,
Dart is 2.19.0
Visual Studio Code 1.74.3
VSC Flutter extension v3.58.0
VSC Dart extension v3.58.0

What I noticed is, that only 1 file was affected in the entire project. The file was originally called `Buttons.dart`. After renaming the file in the VSC Explorer pane to `buttons.dart` the file name changed in the Explorer pane but remained the same on tab in the Editor, and the linter kept complaining.

I managed to fix the issue by inserting an underscore at a random location in the file name, like `button_s.dart`. After this the warning disappeared. I then closed VSC, then reopened it, and renamed the file again back to `buttons.dart`.  And the warning did not come back.

**Comment by DanTup** [comment]
> I managed to fix the issue by inserting an underscore at a random location in the file name, like `button_s.dart`. After this the warning disappeared. I then closed VSC, then reopened it, and renamed the file again back to `buttons.dart`. And the warning did not come back.

Ah, that's good to know. My guess is when you make a change to the filename that is more than just casing, then VS Code _does_ change the internal model, which removes the old casing. Renaming _only_ by the case likely does not. If it works reliably, this might be a better workaround than using `normalizeFileCasing` since that introduces a small performance penalty.

**Comment by sanba-anass** [comment]
issue presist it's annoying since i'm using code gen with mobx and i'm getting an error saying unmatching file of 'part' and 'part of ' even i f i rename it to the correct file naming convention so odd

**Comment by BlarCkat** [comment]
@sanba-anass can you take a screenshot?

**Comment by DanTup** [comment]
@sanba-anass that sounds like a different issue, but it's hard to diagnose without seeing the code. Feel free to open a new issue including a sample project or screenshots that show both of the files in question and the error you're seeing and I'll try to help. Thanks!

**Comment by sanba-anass** [comment]
The file name 'sign_In_page.dart' isn't a lower_case_with_underscores identifier.
Try changing the name to follow the lower_case_with_underscores style.
You may need to close the file and restart VS Code after renaming a file by only casing.

till this day it still the same problem when i fixed the issue the linter still complains about it

**Comment by larssn** [comment]
> I managed to fix the issue by inserting an underscore at a random location in the file name, like `button_s.dart`. After this the warning disappeared. I then closed VSC, then reopened it, and renamed the file again back to `buttons.dart`. And the warning did not come back.

FYI this issue still exists, but managed to fix it with the above workaround.

**Comment by hosmoon** [comment]
> 저도 같은 문제가 있었습니다.
> 
> 내 버전은 다음과 같습니다: Flutter 3.7.0, Dart는 2.19.0 Visual Studio Code 1.74.3 VSC Flutter 확장 프로그램 v3.58.0 VSC Dart 확장 프로그램 v3.58.0
> 
> 제가 발견한 것은 전체 프로젝트에서 단 하나의 파일만 영향을 받았다는 것입니다. 원래 이름은 `Buttons.dart`. VSC 탐색기 창에서 파일 이름을 . 로 변경한 후 탐색기 `buttons.dart`창에서는 파일 이름이 변경되었지만 편집기 탭에는 그대로 남아 있었고, 린터는 계속 문제를 일으켰습니다.
> 
> 파일 이름에 임의의 위치(예: .)에 밑줄을 삽입하여 문제를 해결했습니다 `button_s.dart`. 그 후 경고가 사라졌습니다. VSC를 닫았다가 다시 열어 파일 이름을 .으로 다시 변경했습니다 `buttons.dart`. 그러자 경고가 다시 나타나지 않았습니다.

I love you.
I solve that problem as add _ method.
Thanks!
